# Восстановление золота из руды

Необходимо подготовить прототип модели машинного обучения для компании «Цифры», которая разрабатывает решения для эффективной работы промышленных предприятий.
Модель должна предсказать коэффициент восстановления золота из золотосодержащей руды. У нас есть данные с параметрами добычи и очистки.
Модель поможет оптимизировать производство, чтобы не запускать предприятие с убыточными характеристиками.

В процессе выполнения проекта необходимо:

- Подготовить данные;
- Провести исследовательский анализ данных;
- Построить и обучить модель.

***Технологический процесс***

Когда добытая руда проходит первичную обработку, получается дроблёная смесь. Её отправляют на флотацию (обогащение) и двухэтапную очистку.


**Стадии:**

1. Флотация

Во флотационную установку подаётся смесь золотосодержащей руды. После обогащения получается черновой концентрат и «отвальные хвосты», то есть остатки продукта с низкой концентрацией ценных металлов.
На стабильность этого процесса влияет непостоянное и неоптимальное физико-химическое состояние флотационной пульпы (смеси твёрдых частиц и жидкости).

2. Очистка

Черновой концентрат проходит две очистки. На выходе получается финальный концентрат и новые отвальные хвосты.


***Описание данных***

**Технологический процесс**

* `Rougher feed` — исходное сырье
* `Rougher additions` (или `reagent additions`) — флотационные реагенты: Xanthate, Sulphate, Depressant
    * `Xanthat **` — ксантогенат (промотер, или активатор флотации);
    * `Sulphate` — сульфат (на данном производстве сульфид натрия);
    * `Depressant` — депрессант (силикат натрия).
* `Rougher process` (англ. «грубый процесс») — флотация
* `Rougher tails` — отвальные хвосты
* `Float banks` — флотационная установка
* `Cleaner process` — очистка
* `Rougher Au` — черновой концентрат золота
* `Final Au` — финальный концентрат золота

**Параметры этапов**

* `air amount` — объём воздуха
* `fluid levels` — уровень жидкости
* `feed size` — размер гранул сырья
* `feed rate` — скорость подачи


Нужно спрогнозировать сразу две величины:

1. эффективность обогащения чернового концентрата `rougher.output.recovery`;
2. эффективность обогащения финального концентрата `final.output.recovery`.


Данные находятся в трёх файлах:

* `gold_recovery_train_new_8ds.csv` — обучающая выборка;
* `gold_recovery_test_new_8ds.csv` — тестовая выборка;
* `gold_recovery_full_new_8ds.csv` — исходные данные.

Данные индексируются датой и временем получения информации (признак `date`). Соседние по времени параметры часто похожи.

Некоторые параметры недоступны, потому что замеряются и/или рассчитываются значительно позже. Из-за этого в тестовой выборке отсутствуют некоторые признаки, которые могут быть в обучающей. Также в тестовом наборе нет целевых признаков.

Исходный датасет содержит обучающую и тестовую выборки со всеми признаками.

Данные сырые. Необхоима их проверка и предобработка.

## Загрузка и подготовка данных

In [2]:
from IPython.core.display import display, HTML
display(HTML("<style>.container {width:90% !important; }</style>"))

***Выгрузка библиотек***

In [29]:
import pandas as pd
pd.set_option('display.max_columns', None)

***Константы***

In [4]:
RANDOM_SEED = 0

In [5]:
# Based on: https://stackoverflow.com/a/17303428 (для красивого оформления текста)
class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

***Функции***

In [6]:
#Рандоманая выгрузка строк
show_sample = lambda x: display(x.sample(5, random_state=RANDOM_SEED).reset_index(drop=True))

In [26]:
def com_info(data):
    name = [x for x in globals() if globals()[x] is data][0]
    print(color.UNDERLINE + 'Общая информация по данным '+ name + color.END)
    show_sample(data)
    print(data.info())
    display(pd.DataFrame(round((data.isna().mean()), 2)).style.background_gradient('coolwarm')\
                                                                .set_caption('% пропусков').format({0:'{:.2%}'.format}))
    display(data.describe().T.style.set_caption('Статистика'))
    print('Количество дублей: ', data.duplicated().sum())
    print('******************************************************************')
    print('')

***Загрука данных***

In [27]:
df_train = pd.read_csv('../datasets/gold_recovery_train_new_8ds.csv')
df_test = pd.read_csv('../datasets/gold_recovery_test_new_8ds.csv')
df_full = pd.read_csv('../datasets/gold_recovery_full_new_8ds.csv')

## Изучение общей информации

In [28]:
df_list = [df_train, df_test, df_full]
for i in df_list:
    com_info(i)

Общая информация по данным df_train


date  final.output.concentrate_ag  \
0  2018-06-08 12:59:59                     4.068389   
1  2016-06-24 08:59:59                     4.672297   
2  2017-03-12 03:59:59                     4.999209   
3  2016-06-08 04:59:59                     3.879375   
4  2016-08-01 14:59:59                     4.403698   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                    11.448127                      5.983031   
1                    10.013952                      8.417600   
2                     8.771206                      8.396900   
3                     9.816033                      7.747834   
4                    10.823148                      9.911247   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    45.728369              63.185926              7.436988   
1                    45.142223              70.512351              8.217110   
2                    48.307998              66.501485              9.825832   
3                    47.631350              49.304890              8.156908   
4                    43.927370              69.502636              9.344400   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              2.577877              10.211517              2.042844  ...   
1              1.566327              11.871072              2.433252  ...   
2              3.145526               7.156999              3.558258  ...   
3              2.203170               9.323327              3.720784  ...   
4              2.947228              12.024562              3.293919  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 30.026765   
1                                 15.997023   
2                                 25.012867   
3                                  8.983344   
4                                 20.074004   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -485.043326   
1                                 -499.721909   
2                                 -400.070250   
3                                 -550.003505   
4                                 -401.480443   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 22.022924   
1                                  7.967451   
2                                 22.977935   
3                                  7.115827   
4                                 10.031345   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -466.815927   
1                                 -501.191492   
2                                 -400.084529   
3                                 -550.971423   
4                                 -402.416679   

   secondary_cleaner.state.floatbank5_a_air  \
0                                 25.026081   
1                                 16.024186   
2                                 20.060910   
3                                  7.866452   
4                                 10.010939   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -495.662399   
1                                 -500.172937   
2                                 -449.437364   
3                                 -551.177450   
4                                 -400.062267   

   secondary_cleaner.state.floatbank5_b_air  \
0                                 21.919914   
1                                  7.020786   
2                                 19.983941   
3                                  6.040703   
4                                  4.957975   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.075170   
1                                 -500.459658   
2                                 -449.840148   
3                                 -550.015381   
4                                 -400.499557   

   secondary_cleaner.state.floatbank6_a_air  \
0    

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14149 entries, 0 to 14148
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                14149 non-null  object 
 1   final.output.concentrate_ag                         14148 non-null  float64
 2   final.output.concentrate_pb                         14148 non-null  float64
 3   final.output.concentrate_sol                        13938 non-null  float64
 4   final.output.concentrate_au                         14149 non-null  float64
 5   final.output.recovery                               14149 non-null  float64
 6   final.output.tail_ag                                14149 non-null  float64
 7   final.output.tail_pb                                14049 non-null  float64
 8   final.output.tail_sol                               14144 non-null  float64


Количество дублей:  0
******************************************************************

Общая информация по данным df_test


date  primary_cleaner.input.sulfate  \
0  2016-09-02 05:59:59                     126.329123   
1  2016-10-02 01:59:59                     137.116643   
2  2016-10-05 16:59:59                     136.160058   
3  2017-10-31 11:59:59                     220.706679   
4  2017-12-25 11:59:59                     212.717029   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                          8.003034                         7.160000   
1                          3.972800                         8.600000   
2                          4.001874                         7.700000   
3                          9.139982                         7.040000   
4                         15.009052                         8.150001   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        0.699695                             1400.760536   
1                        0.803511                             1598.994956   
2                        0.701256                             1648.213302   
3                        2.364929                             1700.289848   
4                        0.859222                             1601.389637   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.200817   
1                               -500.061922   
2                               -500.612006   
3                               -508.068208   
4                               -399.591203   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.437375   
1                             1603.985367   
2                             1649.936185   
3                             1700.434596   
4                             1602.644217   

   primary_cleaner.state.floatbank8_b_level  \
0                               -500.679938   
1                               -499.302443   
2                               -499.849197   
3                               -489.509097   
4                               -397.938382   

   primary_cleaner.state.floatbank8_c_air  ...  \
0                             1399.429209  ...   
1                             1600.063771  ...   
2                             1647.739364  ...   
3                             1542.893209  ...   
4                             1301.027142  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 11.960962   
1                                 12.015829   
2                                 13.988463   
3                                 18.019398   
4                                 19.997723   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -500.130469   
1                                 -580.205684   
2                                 -498.835101   
3                                 -502.523088   
4                                 -498.411588   

   secondary_cleaner.state.floatbank4_b_air  \
0                                  7.924387   
1                                  9.996712   
2                                 13.073075   
3                                 15.929214   
4                                 14.901334   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -500.362897   
1                                 -499.568551   
2                                 -501.503654   
3                                 -501.441100   
4                                 -498.971556   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  7.849120   
1                                 10.029440   
2                                 10.127322   
3                                 14.962390   
4                                 11.040440   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -500.357611   
1                                 -500.834429   
2                                 -498.489453   
3     

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        5290 non-null   object 
 1   primary_cleaner.input.sulfate               5286 non-null   float64
 2   primary_cleaner.input.depressant            5285 non-null   float64
 3   primary_cleaner.input.feed_size             5290 non-null   float64
 4   primary_cleaner.input.xanthate              5286 non-null   float64
 5   primary_cleaner.state.floatbank8_a_air      5290 non-null   float64
 6   primary_cleaner.state.floatbank8_a_level    5290 non-null   float64
 7   primary_cleaner.state.floatbank8_b_air      5290 non-null   float64
 8   primary_cleaner.state.floatbank8_b_level    5290 non-null   float64
 9   primary_cleaner.state.floatbank8_c_air      5290 non-null   float64
 10  primary_clea

Количество дублей:  0
******************************************************************

Общая информация по данным df_full


date  final.output.concentrate_ag  \
0  2017-02-08 14:59:59                     5.295988   
1  2016-10-29 13:59:59                     3.020476   
2  2017-08-10 06:59:59                     3.455687   
3  2018-03-20 17:59:59                     5.360341   
4  2017-02-26 19:59:59                     4.645203   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     8.195301                     13.379583   
1                     9.071122                     16.816693   
2                    12.662607                      6.890618   
3                    12.297913                      6.589168   
4                    10.322894                     14.283728   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    47.758663              65.280176              9.925318   
1                    49.797229              63.724514              8.165338   
2                    45.799572              62.002449              7.103502   
3                    42.586163              56.365844              8.325812   
4                    46.100397              74.086821             12.162428   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              3.387088               7.605304              3.382011  ...   
1              3.167601              10.705567              3.591239  ...   
2              4.485478              11.579938              3.446024  ...   
3              1.211108               8.217830              2.221499  ...   
4              4.429356               9.392597              3.720634  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 24.937716   
1                                 19.999104   
2                                 16.031300   
3                                 23.051530   
4                                 24.994581   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -499.827034   
1                                 -500.526123   
2                                 -500.548355   
3                                 -500.327972   
4                                 -398.620115   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 22.955855   
1                                 16.899111   
2                                 12.014055   
3                                 14.984485   
4                                 23.010430   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -499.981726   
1                                 -500.418689   
2                                 -399.804953   
3                                 -499.893529   
4                                 -399.770903   

   secondary_cleaner.state.floatbank5_a_air  \
0                                 23.018173   
1                                 17.000241   
2                                 15.987138   
3                                 30.004056   
4                                 22.975057   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -499.465128   
1                                 -497.845707   
2                                 -500.414523   
3                                 -499.853016   
4                                 -455.682049   

   secondary_cleaner.state.floatbank5_b_air  \
0                                 20.014626   
1                                 13.990843   
2                                 11.990190   
3                                 15.031900   
4                                 20.005831   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.003639   
1                                 -499.834974   
2                                 -500.074479   
3                                 -500.068586   
4                                 -449.548036   

   secondary_cleaner.state.floatbank6_a_air  \
0    

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19439 entries, 0 to 19438
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                19439 non-null  object 
 1   final.output.concentrate_ag                         19438 non-null  float64
 2   final.output.concentrate_pb                         19438 non-null  float64
 3   final.output.concentrate_sol                        19228 non-null  float64
 4   final.output.concentrate_au                         19439 non-null  float64
 5   final.output.recovery                               19439 non-null  float64
 6   final.output.tail_ag                                19438 non-null  float64
 7   final.output.tail_pb                                19338 non-null  float64
 8   final.output.tail_sol                               19433 non-null  float64


Количество дублей:  0
******************************************************************



***df_train***

- Типы данных не оптимизированы, необходимо понизить разрядность признаков
- В данных есть пропуски, больше всего в признаке `secondary_cleaner.output.tail_sol` (11%)
- Некоторые признаки отрицательные - например, `primary_cleaner.state.floatbank8_b_level`, `primary_cleaner.state.floatbank8_c_level`, `rougher.state.floatbank10_a_level` и т.д.
- Полных дублей в данных нет.

## Предобработка